### DISCUSSION TITLE
Update an array of objects through rpc

### DISCUSSION THREAD

**Discussion by pierroo** [supabase/supabase#16817] [Questions] [answered]
Hi,

I am banging my head trying to update a column by inserting (ideally, if not already exists) an object (json) into an array of object.

The equivalent of addToField in mongodb, or a basic push in node.js, but it's killing me in those rpc function

Here is what I tried so far:

```
create or replace function pick_answer(msg_id int, type text)
returns void
as
$func$
    update room_messages
    set participants = array_append(participants, $2)
    where id = $1
$func$
language sql;
```

But I keep getting the error "Failed to run sql query: function array_append(json[], text) does not exist"

Ideally instead of $2, I would even like to have array_append(participants, {id: auth.uid(), type: $2} )

but obviously that's not how sql works and I can't have {} in there;

My "participants" column is a type json{} with the "define as array" checked in, and ideally should be of the form: 
[
{id: uuid, type: text},
...
]

any help would be appreciated! :)

**Comment by GaryAustin1** [comment, marked as answer]
Do you want an array column with lots of different json objects or a json object with an array of objects?
This seems to show updating a json object that contains an array:
https://stackoverflow.com/questions/41593815/update-json-array-in-postgres (goto fixing your typos).
Another example: https://kevcodez.de/posts/2020-09-13-postgres-jsonb-update-object-array/

If you truly want a jsonb[] column of arrays of json then array_append is I believe the correct approach.
You would need your 2nd item to be json is likely what it is complaining about.  You probably need `jsonb_build_object` here for that 2nd parameter https://www.postgresql.org/docs/9.5/functions-json.html

I've never used a jsonb[] array column though so can't confirm.

**Reply by pierroo** [reply]
What I am looking for is an array column with lots of different json objects (I guess?)

basically I want that columns to contain for example:

[
{id: uuid, type: text},
...
]

isn't that an array column with many objects inside?

I really don't see what would be the point of a json object containing an array of object; but maybe it's me coming from firebase being super confused about those columns types?

**Reply by GaryAustin1** [reply]
I only use jsonb columns myself.  They return a json object you can manipulate all the json it in the client.  JSON can contain arrays, text, numbers etc as objects.

If you are doing array of json, then each item returned in a request is a separate json item and you manipulate the json in them after you get them from the array.   Your row returned as a field that is an array.  Inside that array are bunch of separate json objects.

Not sure I can be more help.

**Reply by GaryAustin1** [reply]
There should be lots of info on the web for Postgres and json/jsonb and jsonb[]/json[] columns.  It is not specific to Supabase.

**Reply by pierroo** [reply]
I fully understand the difference between json containing arrays of whatever, and arrays of json;
and what I need is definitely array of json as displayed in this screenshot:
![image](https://github.com/supabase/supabase/assets/17619412/8a0d1174-34db-4976-9f8c-16034074adf6)

so in order to append the new json to the array, it seems that as per your suggestion I was needing:
json_build_object('id', auth.uid(),'type',$2)

the SQL to create the function now runs properly, thank you so much @GaryAustin1 !

**Reply by pierroo** [reply]
But I unfortunately just realized it doesn't care about duplicate append;
I can have many times the same object inside the array which is not what I want, I guess that's how array_append work, it's just like an array.push() in JS

I was searching online, found some "DISTINCT" use, but they are all about array of text or integers, not array of objects where I would make sure there isn't already an object with the same id value as auth.uid();

is there any magic function that would take care of that?

**Reply by GaryAustin1** [reply]
Just google and see.  I don't know.  You may have to check if it is there first and then not insert it.
I don't think push in node.js cares about unique either.  I don't know MongoDB.

Are you sure this structure you are doing should be in a single row and column?  Relational databases normally work better if you have a table to represent your data, then you can also have `unique` enforced on column values for a row.  Trivially to add/delete and search.

**Reply by pierroo** [reply]
I just asked chatgpt to be honest ^^
and it told me I should indeed first check if exist, then only do the update like that:

```
CREATE OR REPLACE FUNCTION pick_shifumi(msg_id bigint, type text)
RETURNS VOID
AS
$func$
BEGIN
    -- Check if the JSON object already exists in the array
    IF NOT EXISTS (
        SELECT 1
        FROM room_messages
        WHERE id = $1
        AND participants @> ARRAY[json_build_object('type', $2)::jsonb]
    ) THEN
        UPDATE room_messages
        SET participants = array_append(participants, json_build_object('id', auth.uid(), 'type', $2))
        WHERE id = $1;
    END IF;
END
$func$
LANGUAGE plpgsql;
```
(I haven't tested / adjusted it yet)

as for the structure, it is very likely I am not doing it right and am still being brain washed by all these years of firebase / nosql.
The context is this one:

I have a chat room.
People send chat messages I am subscribing too from the front.
I am subscribing to insert, obviously, but also to update;
because some messages are like a "shifumi", where many can participate by clicking on the message
=> everyone participating, should then be appended as an object (containing both their uid and their answer to shifumi) to that participants column in the actual message row based on its msg_id
=> so from the front, when someone participates, I can update the view of that chat bubble accordingly realtime

I have no clue how or why I would do that in any other way but to have it directly inside the message itself?

**Reply by GaryAustin1** [reply]
You would normally do that with a join table.  The join table would have a message foreign key and user profile foreign key.  You would add a pair to that every time a user adds their name.  Then you would just select the message and join the profile table to get the array of user profile info.

BUT realtime can't join.  So your choice is embed the info in the row like you are doing or when you get the payload with the message you make a call back to the server to get the message with the join table in the select.

**Reply by pierroo** [reply]
That's exactly the problem; since realtime can't join in so many occasions we have no other choice but to embed it;
I'm really not a fan of having to do extra server calls on top of a realtime that was supposed to handle it in the first place

In any case, I obviously have a lot to learn about postgres if I want to make further progress, thank you as always for your time and help :)

**Comment by pierroo** [comment]
Just in case anyone faces the same issue / question, here is what ended up working for me to make sure I'm not pushing a duplicate "id" in my array of json:

```
CREATE OR REPLACE FUNCTION pick_shifumi_test(msg_id bigint, type text)
RETURNS VOID
AS
$func$
DECLARE
    existing_id_found BOOLEAN;
    participant json;
BEGIN
    -- Check if the "id" already exists in any object inside the array
    existing_id_found := EXISTS (
        SELECT 1
        FROM room_messages
        WHERE id = $1
        AND EXISTS (
            SELECT 1
            FROM unnest(participants) AS p
            WHERE (p->>'id')::uuid = auth.uid()
        )
    );

    IF NOT existing_id_found THEN
        participant := json_build_object('id', auth.uid(), 'type', type)::json;
        UPDATE room_messages
        SET participants = array_append(participants, participant)
        WHERE id = $1;
    END IF;
END
$func$
LANGUAGE plpgsql;

```
